das: Parallelise DASer

walldiss opened the pull request:
Resolves 
* https://github.com/celestiaorg/celestia-node/issues/483
* https://github.com/celestiaorg/celestia-node/issues/504
* https://github.com/celestiaorg/celestia-node/issues/793
* https://github.com/celestiaorg/celestia-node/issues/1049

Distribute data availability sampling across pool of workers.

New DASer implementation bring pool of workers. Core features:
 - Sampling from header with height 1 to latest known one.
 - Subscription to new headers  is running in parallel and pushes new headers to DASer callback on demand.
 - New found headers are processed with higher priority, but from only 1 worker at time. This way catch-up workers are not block by low speed of recent headers sampling.
 - Add very basic error handling: If header is impossible to sample its height is stored for retry upon restart. Amount of retries persist over checkpoint until header is sampled successfully
 - Store sampling state periodically in case of force quit.
 
 TODO:
  - add tracing
  - add metrics

renaynay commented on das/daser.go:
```diff
@@ -135,125 +178,151 @@ func (d *DASer) sample(ctx context.Context, sub header.Subscription, checkpoint
+	// store latest DASed checkpoint to disk here to ensure that if DASer is not yet
+	// fully caught up to network head, it will resume DASing from this checkpoint
+	// up to current network head
+	// TODO @renaynay: Implement Share Cache #180 to ensure no duplicate DASing over same
+	//  header
```
todo can finally be removed

walldiss replied:
todo no more

renaynay commented on das/state.go:
```diff
@@ -28,6 +28,9 @@ type RoutineState struct {
 	// height of the routine
 	LatestSampledSquareWidth uint64 `json:"latest_sampled_square_width"`
+	// amount of sampled headers
+	SampledHeadersAmount uint64 `json:"sampled_header_amount"`
```
we don't need to track this on daser IMO unless it's for metrics purposes, but i'm sure tracing will provide a better mechanism for this.

distractedm1nd commented on das/stats.go:
```diff
@@ -0,0 +1,59 @@
+// RoutineState contains important information about the state of a
+// current sampling routine.
+type RoutineState struct {
+	// reports if an error has occurred during the routine's
```
You could add the struct fields' names to the beginning of the comments so they are all godoc'ed

distractedm1nd commented on das/checkpoint_store.go:
```diff
@@ -13,35 +13,45 @@ var (
 }
 
-// loadCheckpoint loads the DAS checkpoint height from disk and returns it.
+// loadCheckpoint loads the DAS checkpoint height maxKnown disk and returns it.
```
I assume this was a search and replace error

walldiss replied:
IDE renamed commets with variable for some reason. Sorry for that

distractedm1nd commented on das/checkpoint_store.go:
```diff
@@ -13,35 +13,45 @@ var (
+}
 
-	return ds.Put(ctx, checkpointKey, buf)
+// storeCheckpoint stores the given DAS checkpoint maxKnown disk.
```
search and replace error (to -> maxKnown)

distractedm1nd commented on das/manager.go:
```diff
@@ -0,0 +1,175 @@
+}
+
+// BackgroundStorer periodically stores state to keep stored version up-to-date in case force quit happen
+func (sm *samplingManager) runBackgroundStorer(ctx context.Context, interval time.Duration) {
```
I would replace all instances of "storer" with "store"

distractedm1nd reviewed (commented):
just looking at docs and names for now

renaynay reviewed (changes requested):
Leaving an intermediate review for now -- 

I like the general direction. I think there are things that could be simplified (e.g. the backgroundStore). Also, can we rename `discovery` of headers to something other than discovery? Just something that indicates it's recent/received through subscription rather than some active discovery process.

Another thing is the frequency/duplicity of storing the checkpoint to disk:


```
2022-08-18T10:34:56.270+0200	INFO	das	das/daser.go:200	discovered new header	{"height": 181780}
2022-08-18T10:34:56.270+0200	INFO	das	das/state.go:63	added recent headers maxKnown DASer priority queue 	{"from_height": 181779, "to_height": 181780}
2022-08-18T10:35:15.924+0200	INFO	das	das/daser.go:151	stored checkpoint to disk	{"checkpoint": {"min_sampled_height":173870,"max_known_height":181780,"skipped":{}}}
2022-08-18T10:36:15.920+0200	INFO	das	das/daser.go:151	stored checkpoint to disk	{"checkpoint": {"min_sampled_height":173870,"max_known_height":181780,"skipped":{}}}
2022-08-18T10:37:15.918+0200	INFO	das	das/daser.go:151	stored checkpoint to disk	{"checkpoint": {"min_sampled_height":173870,"max_known_height":181780,"skipped":{}}}
2022-08-18T10:38:15.921+0200	INFO	das	das/daser.go:151	stored checkpoint to disk	{"checkpoint": {"min_sampled_height":173870,"max_known_height":181780,"skipped":{}}}
2022-08-18T10:39:15.916+0200	INFO	das	das/daser.go:151	stored checkpoint to disk	{"checkpoint": {"min_sampled_height":173870,"max_known_height":181780,"skipped":{}}}
2022-08-18T10:40:15.915+0200	INFO	das	das/daser.go:151	stored checkpoint to disk	{"checkpoint": {"min_sampled_height":173870,"max_known_height":181780,"skipped":{}}}
```

I am running your branch against mamaki and this is taken from my logs ^. 

I think this PR would benefit from some further simplification as well as cleaning up of weird behaviours like the one I just mentioned. Also more documentation :) There is a `doc.go` file and it would be nice if you could update it.

renaynay commented:
Also, when stopping the node (without force quitting), I get the following: 

```
2022-08-18T10:45:07.166+0200	ERROR	node	node/node.go:142	Stopping Light Node: cannot close topic: outstanding event handlers or subscriptions; cannot close topic: outstanding event handlers or subscriptions
Error: cannot close topic: outstanding event handlers or subscriptions; cannot close topic: outstanding event handlers or subscriptions
```

Wondertan commented on das/checkpoint_store.go:
```diff
@@ -13,35 +13,45 @@ var (
 )
 
+type checkPoint struct {
+	MinSampled uint64         `json:"min_sampled_height"` // lowest sampled height
+	MaxKnown   uint64         `json:"max_known_height"`   // height of the newest known header
+	Skipped    map[uint64]int `json:"skipped"`
```
They are because of the JSON marshaling.

renaynay replied:
Aha correct.

Wondertan commented on das/checkpoint_store.go:
```diff
@@ -13,35 +13,45 @@ var (
-func loadCheckpoint(ctx context.Context, ds datastore.Datastore) (int64, error) {
-	checkpoint, err := ds.Get(ctx, checkpointKey)
+func loadCheckpoint(ctx context.Context, ds datastore.Datastore) (checkPoint, error) {
+	bs, err := ds.Get(ctx, checkpointKey)
```
Typo probably

walldiss replied:
byte slice abbreviation. Can also rename it to `data`

walldiss commented on das/checkpoint_store.go:
```diff
@@ -13,35 +13,45 @@ var (
 )
 
+type checkPoint struct {
+	MinSampled uint64         `json:"min_sampled_height"` // lowest sampled height
+	MaxKnown   uint64         `json:"max_known_height"`   // height of the newest known header
+	Skipped    map[uint64]int `json:"skipped"`
```
Value represents amount of retries for given height. Sorry for confusion, I will add description to `state` struct

walldiss commented on das/manager.go:
```diff
@@ -0,0 +1,175 @@
+	cancel          context.CancelFunc
+}
+
+func newSamplingManager(concurrency int,
```
It needs to be in constructor because different tests use diferent concurrency values and can run in parallel. On the other hand for application it takes value from pkg level constant that defined `daser.go`. 

BTW it is not exposed to user but could be put in config in future. Concurrency level could affect sampling speed on cost or machine resources consumption so could subject of user configuration.

renaynay replied:
I think we want to limit user configuration for this kind of thing at the moment, but definitely in the future would consider this.

walldiss commented on das/checkpoint_store.go:
```diff
@@ -13,35 +13,45 @@ var (
 // wrapCheckpointStore wraps the given datastore.Datastore with the `das`
-// prefix. The checkpoint store stores/loads the DASer's checkpoint to/from
-// disk using the checkpointKey. The checkpoint is stored as a uint64
+// prefix. The checkpoint store stores/loads the DASer's checkpoint maxKnown/maxKnown
```
fixed. Thank you

walldiss commented on das/checkpoint_store.go:
```diff
@@ -13,35 +13,45 @@ var (
 }
 
-// loadCheckpoint loads the DAS checkpoint height from disk and returns it.
+// loadCheckpoint loads the DAS checkpoint height maxKnown disk and returns it.
```
same :)

walldiss commented on das/checkpoint_store.go:
```diff
@@ -13,35 +13,45 @@ var (
 	checkpointKey = datastore.NewKey("checkpoint")
 )
 
+type checkPoint struct {
```
ty, renamed

walldiss commented on das/checkpoint_store.go:
```diff
@@ -13,35 +13,45 @@ var (
 		if err == datastore.ErrNotFound {
-			log.Debug("checkpoint not found, starting sampling at block height 1")
-			return 0, nil
+			log.Debug("checkpoint not found, starting sampling at lock height 1")
```
yep. fixed

walldiss commented on das/checkpoint_store.go:
```diff
@@ -13,35 +13,45 @@ var (
-		// if no checkpoint was found, return checkpoint as
-		// 0 since DASer begins sampling on checkpoint+1
+		// if no checkpoint was found, return checkpoint as 0
+		// DASer begins sampling on discovering new header
```
rephrased comment and moved it from place in code where checkpoint is loaded to place where it is handled

Wondertan commented on das/daser.go:
```diff
@@ -17,25 +18,39 @@ import (
+	concurrency        = 256
+	jobsBufferSize     = 16
+	storeInterval      = 10 * time.Minute
+)
```
Should we make it private? It can't be used outside really and pollutes Public API

renaynay replied:
What are you pointing to? The `Config`?

Wondertan replied:
Yes, I guess this is the issue with WEB VScode I've used for reviewes. Its buggy unfortunately

Wondertan commented on das/checkpoint_store.go:
```diff
@@ -2,7 +2,7 @@ package das
 
```
We might want to rename this file to just `checkpoint`

walldiss replied:
renamed. Thank you

Wondertan commented on das/checkpoint_store.go:
```diff
@@ -13,35 +13,44 @@ var (
+
+	cp := checkpoint{}
+	err = json.Unmarshal(bs, &cp)
+	return cp, err
 }
 
```
If we always return zero value here, we might return `json.Unmarshal(bs, &cp)` diretly.

Wondertan commented on das/manager.go:
```diff
@@ -0,0 +1,187 @@
+
+	var prevMinSampled uint64
+	for {
+		// blocked by ticker to perform store only once in period of time
```
Why is this called `updateStates` if we only manage done channel there?

Wondertan commented on das/manager.go:
```diff
@@ -0,0 +1,187 @@
+
+func (sm *samplingManager) runCoordinator(ctx context.Context) {
+	jobsCh := sm.jobsCh
+	noop := make(chan uint64)
```
I've only used nil in such cases before and have never thought that one can just use unbuffered chan. Wondering what would be the most readable

renaynay replied:
Please less channels.

Wondertan commented on das/manager.go:
```diff
@@ -0,0 +1,187 @@
+			}
+		case res := <-sm.resultCh:
+			sm.state.handleResult(res)
+		case sm.storeCh <- *sm.state:
```
I am a bit concerned about the copy here and wondering if it gets copied on each loop/try to write to `sm.storeCh` even when its blocked. Basically, there might be some idle copies that we should avoid, but this has to be verified.

Wondertan replied:
We can investigate this or just interpolate the storing ticker into the runCoordinator. There is no big reason to keep it separated

Wondertan replied:
Considering that state is only used inside samplingManager, you can store it as value(without asterisk) and writing ot the storeCh will do copying automatically

walldiss replied:
good point, changed it to be value field instead of pointer

Wondertan commented on das/state.go:
```diff
@@ -1,53 +1,140 @@
-	End   time.Time `json:"end"`
-	Error error     `json:"error"`
+func (s *state) updateMaxKnown(last uint64) bool {
+	// seen this header before
+	if last == s.maxKnown {
+		return false
+	}
```
Did you see this case in practice or do you assume that it might happen? AFAIU, this case is impossible and done on HeaderGetter level, if it does happen, then there is a bug

Wondertan commented on das/state.go:
```diff
@@ -1,53 +1,140 @@
 
-	catchUpLk sync.RWMutex
-	catchUp   JobInfo // tracks information related to the `catchUp` routine
+	priorityBusy bool // semaphore to allow only one priority item to be processed at time
```
We don't need to implement the solution for broken Mamaki issues. Details are dicussed in sync

Wondertan commented on das/daser.go:
```diff
@@ -17,25 +18,39 @@ import (
-	cancel context.CancelFunc
+const (
+	priorityBufferSize = 1024
+	concurrency        = 256
```
I really like the process how you measured and choose this value. Can you pls document it?

Wondertan commented:
Oops, sorry for making each comment a separate review. I've started using different software for reviews and not yet got used to it

Wondertan commented:
As per the sync discussion, we decided to rework one by one header height handling in favor of height ranges

Wondertan commented:
TODO list from sync call
* backgroundStore finish
* test documentation
* DASer state API
* Docs

Wondertan commented on das/checkpoint_store.go:
```diff
@@ -13,35 +13,45 @@ var (
 	}
-	return int64(binary.BigEndian.Uint64(checkpoint)), err
+
+	cp := checkpoint{}
+	err = json.Unmarshal(bs, &cp)
+	return cp, err
 }
```
What's the point of passing value? The q is not about this place specifically, but about all the places where the checkpoint is copied when passed.

walldiss replied:
You are right, returning and passing by copy mean extra coping. When returning and passing by copy it will be allocated on short-lived stack, on the other hand returning by pointer will require compiler to do heap allocation.

A stack allocation is short-lived and typically very cheap. No Garbage Collection is required for a stack allocation as the end of the function is also the end of the variable’s lifetime. On the other hand, a heap allocation is long-lived and considerably more expensive. When allocating onto the heap, the runtime must find a contiguous piece of memory where the new variable fits. Additionally, it must be garbage-collected when the variable is no longer us

ed. Both operations are orders of magnitudes more expensive than a stack allocation.

To showcase benefits of passing by copy I've recorder 2 test traces on local machine:
[copy_trace.out.zip](https://github.com/celestiaorg/celestia-node/files/9431324/copy_trace.out.zip)
[pointer_trace.out.zip](https://github.com/celestiaorg/celestia-node/files/9431325/pointer_trace.out.zip)
You can run them to investigate allocation details and GC time by 
```
go tool trace pointer_trace.out
```
There is also small benchmark in `coordinator_test.go` you I've run with `-benchmem` flag. Results for pointer vs copy on my local machine:
```
BenchmarkManager/bench_run-copy10         	 7310666	       145.9 ns/op	      18 B/op	       0 allocs/op
BenchmarkManager/bench_run-pointer-10         	 7417952	       182.2 ns/op	      21 B/op	       0 allocs/op
```

Wondertan replied:
Thanks for the detailed proof. That's what we were missing in our knowledge, and I am pretty sure there are other places throughout the codebase where we could use value rather than ptrs. Pls report those if you find them or even fix them through PRs 🙏🏻

Wondertan commented on das/state.go:
```diff
@@ -1,53 +1,204 @@
+	catchUpDoneCh chan struct{} // blocks until all headers are sampled
+}
+
+func initSamplingState(samplingRangeSize uint64, c checkpoint) state {
```
Same q: why do we copy checkpoint and state values everywhere?

Wondertan commented on das/daser.go:
```diff
@@ -68,195 +72,82 @@ func (d *DASer) Start(context.Context) error {
-			if err == context.Canceled {
-				return
-			}
+	log.Infow("loaded checkpoint", "height", cp)
```
```suggestion
	log.Infow("loaded checkpoint", "last_height", cp.MinSampled)
```

Wondertan replied:
Or we should make prettier `String()` impl for the checkpoint

walldiss replied:
good idea, done it

Wondertan commented on das/daser.go:
```diff
@@ -68,195 +72,82 @@ func (d *DASer) Start(context.Context) error {
-		case <-d.sampleDn:
-		case <-ctx.Done():
-			return ctx.Err()
+			log.Warnw("set max known height from store", "height", cp.MaxKnown)
```
```suggestion
			log.Warnw("checkpoint not found, initializing with height 1")
```

walldiss replied:
+, thank you

Wondertan commented on das/daser.go:
```diff
@@ -68,195 +72,82 @@ func (d *DASer) Start(context.Context) error {
-		}
+	go d.sampler.runCoordinator(runCtx)
+	// subscribe to new headers
+	go d.runSubscriber(runCtx, sub, d.sampler.listen)
```
Feel conflicted about circular dependencies between sampler/daser/state(and storer soon). A bit of a mess, but not a big deal considering them all being private.

walldiss replied:
They same package so shouldn't be a problem. When implementing I've tried to create tree-like structure of dependencies:
DASER
   -coordinator
      -state
      -worker
      -store
   - discovery
      - subscription
      - push callback to coordinator (not a dependancy)
   - background store
      - store
      - pull callback to coordinator (not a dependancy)

Wondertan commented on das/daser.go:
```diff
@@ -68,195 +72,82 @@ func (d *DASer) Start(context.Context) error {
-			// report previous height as the last successfully sampled height
-			return height - 1, err
-		}
+func (d *DASer) fetch(ctx context.Context, height uint64) error {
```
Let's call it `sampleHeight`. Understand where is fethc coming from but it's a bit off

walldiss replied:
ty, renamed for clarity

Wondertan commented on das/daser.go:
```diff
@@ -68,195 +72,82 @@ func (d *DASer) Start(context.Context) error {
-		d.state.catchUpLk.Lock()
-		d.state.catchUp.Height = uint64(h.Height)
-		d.state.catchUpLk.Unlock()
+func (d *DASer) storeCheckpount(ctx context.Context, cp checkpoint) {
```
```suggestion
func (d *DASer) storeCheckpoint(ctx context.Context, cp checkpoint) {
```

Wondertan commented on das/daser.go:
```diff
@@ -68,195 +72,82 @@ func (d *DASer) Start(context.Context) error {
-}
+// Stop stops sampling.
+func (d *DASer) Stop(ctx context.Context) error {
+	if atomic.CompareAndSwapInt32(&d.closed, 0, 1) {
```
stylistic nit: let's keep logic outside of the if case and keep return nil there instead

Wondertan commented on das/manager.go:
```diff
@@ -0,0 +1,190 @@
+	// wait for coordinator to exit and store state
+	select {
+	case <-sm.coordinatorDone:
+		sm.storeState(ctx, sm.state.toCheckPoint())
```
The coordinator itself can do this op. This way we can avoid coordinatorDone and its handling

Wondertan replied:
Well, actually, the `storeState` on stop is more correct to be done by the ctx given to `Stop` as it is right now, so no need to change this

Wondertan replied:
Still, we should store the checkpoint when workers are done

walldiss replied:
Yep, you got it right. `store` function  is called from `Stop()` to use with `stopCtx` instead of `runCtx` of the coordinator.

Added extra call for `store` after workers exit as you suggested. There was a possibility they progressed 1 more header ahead.

Wondertan commented on das/daser.go:
```diff
@@ -285,51 +176,60 @@ func (d *DASer) sampleHeader(ctx context.Context, h *header.ExtendedHeader) erro
-	d.state.sampleLk.Lock()
-	defer d.state.sampleLk.Unlock()
-	d.state.sample.IsRunning = true
+		emit(ctx, uint64(h.Height))
```
Ideally, we would avoid "regetting" the header and use the one we have, but that's not a big deal in this case

walldiss replied:
Do you refer to this lines?
```
func (s *coordinatorState) updateMaxKnown(last uint64) {
	// seen this header before
	if last <= s.maxKnown {
		return
	}
	...
```
This case is added just in case GetHeader() API changes behavior. If it happens DASer will be screwed without this case.

I've added log WARN to track this specific scenario

Wondertan replied:
No, I meant the subscription gives the actual Header while we just ignore it and then get it again via HeaderGetter.

walldiss replied:
Ah, I see. It means, there are will be extra call for header store for new headers. Not ideal, but should be fast considering headers cache

Wondertan replied:
Yes

Wondertan commented on das/daser.go:
```diff
@@ -17,25 +18,28 @@ import (
 var log = logging.Logger("das")
 
+const (
+	samplingRange = 1
```
We should increase the value here. It feels like it should be close to the number of headers we retrieve in the batch via HeaderEx, but this is better to be analyzed.

walldiss replied:
Increased to 64, further tweaking performance testing is needed to find optimal values

Wondertan commented on das/worker.go:
```diff
@@ -0,0 +1,103 @@
+	if to > max {
+		to = max
+	}
+	return job{
```
Same here. Lot's of jobs are allocated during runtime, not sure why the needs to be copied while being passed through each method boundary

walldiss replied:
Answered in upper comment

Wondertan reviewed (commented):
Awesome work! The logic is correct AFAICS, and I wasn't able to find any bugs. Left mostly stylistic nits/questions and thoughts on things. Looking forward to the above TODO's changes to land

walldiss commented on das/worker.go:
```diff
@@ -0,0 +1,103 @@
+	defer w.lock.Unlock()
+	if err != nil {
+		w.state.Failed = append(w.state.Failed, curr)
+		w.state.Err = multierror.Append(w.state.Err, fmt.Errorf("height: %v, Err: %w", curr, err))
```
swaped libs to use `uber-go/multierrr`

Wondertan reviewed (commented):
* We can consider moving `lock` and `done` to some utility primitives somewhere in libs.
* We should start making global var params configurable. I don't mind this landing in this PR

Wondertan commented on das/daser_test.go:
```diff
@@ -469,7 +216,7 @@ func (m *mockGetter) fillSubWithHeaders(
 
 		sub.Headers[index] = randHeader
-		// also store to mock getter for duplicate fetching
+		// also store to mock getter for duplicate sampleing
```
```suggestion
		// also store to mock getter for duplicate sampling
```

walldiss commented on das/worker.go:
```diff
@@ -0,0 +1,93 @@
+		err := w.sample(ctx, curr)
+		w.setResult(curr, err)
+
+		select {
+		case <-ctx.Done():
+			return
+		default:
+		}
```
Main reason is to have `curr` iterator at latest attempt before context got canceled.  Further it is stored stored in checkpoint and after restart worker will resume from same `curr`. Otherwise all headers that are left in job after `cancel()` call will be stored as error.

`Done()` brings very little overhead because it is `<-chan struct{}` and should not affect other parallel `Done()` calls

Wondertan replied:
> Done() brings very little overhead because it is <-chan struct{} and should not affect other parallel Done() calls

I meant the mutex inside the Done call

> Main reason is to have curr iterator at latest attempt before context got canceled. Further it is stored stored in checkpoint and after restart worker will resume from same curr. Otherwise all headers that are left in job after cancel() call will be stored as error.

Ok, so there is a correctness issue here.

Wondertan replied:
This makes me think about retrying for failed headers

Wondertan replied:
Which is actually here. This all takes time to absorb and chew

walldiss replied:
Yeah, it is sort of hidden inside save/load checkpoint functionality.

walldiss commented on das/coordinator_test.go:
```diff
@@ -0,0 +1,558 @@
+
+		// wait for coordinator to finish catchup
+		assert.NoError(t, coordinator.state.waitCatchUp(ctx))
+		assert.Emptyf(t, coordinator.state.failed, "failde list should be empty")
```
fixed, thank you

walldiss commented on das/coordinator_test.go:
```diff
@@ -0,0 +1,558 @@
+		coordinator.state = initSamplingState(samplingRange, sampler.checkpoint)
+		go coordinator.run(ctx)
+
+		time.Sleep(time.Second)
```
agree, downscaled to 50ms

walldiss commented on das/stats.go:
```diff
@@ -0,0 +1,30 @@
+	Failed map[uint64]int `json:"failed,omitempty"`
+	// Workers has information about each currently running worker stats
+	Workers []WorkerStats `json:"workers,omitempty"`
+	// Concurrency currently running parallel workers
+	Concurrency int `json:"concurrency"`
```
Hard to count json objects when there are 64 of them, so I thought it wont hurt to have an extra field :)

renaynay replied:
Yes it's definitely hard to count json objects when there's 64 of them. I am in favour of decreasing worker amount to 16 for both readability and debug purposes. Each worker can take a larger range.

Wondertan commented:
Rebase should fix failed integration test

renaynay commented on das/daser.go:
```diff
@@ -17,49 +18,59 @@ import (
 
+// TODO: parameters needs performance testing on real network to define optimal values
+const (
+	samplingRange           = 64
+	defaultConcurrencyLimit = 64
+	storeInterval           = 10 * time.Minute
```
pls doc every const

walldiss replied:
Done

renaynay commented on das/store.go:
```diff
@@ -0,0 +1,114 @@
+	checkpointKey = datastore.NewKey("checkpoint")
+)
+
+type checkpointStore interface {
```
Why does this have to be an interface? Why not use dataStore directly? I understand the need for testing, but there is an in-mem ds that you can use.

walldiss replied:
Yeah interface was mainly introduced to create mock independent of third party package functionality on a cost of extra abstraction. Otherwise would be store->load->assert semantic. But tests are less important then main code, I think it is good idea to get rid of it. Thank you!

renaynay commented on das/store.go:
```diff
@@ -0,0 +1,114 @@
+}
+
+// wrapCheckpointStore wraps the given datastore.Datastore with the `das`
+// prefix. The checkpoint store stores/loads the DASer's checkpoint to/from
+// disk using the checkpointKey. The checkpoint is stored as an uint64
+// representation of the height of the latest successfully DASed header.
```
this is no longer true.

walldiss replied:
fixed

renaynay commented on das/store.go:
```diff
@@ -0,0 +1,114 @@
+	return &dataStore{namespace.Wrap(ds, storePrefix)}
+}
+
+// loadCheckpoint loads the DAS checkpoint from disk and returns it.
+// If there is no known checkpoint, it returns height 0.
```
please godoc correctly

walldiss replied:
fixed

renaynay commented on das/store.go:
```diff
@@ -0,0 +1,114 @@
+			log.Debug("checkpoint not found, starting sampling at block height 1")
+			return checkpoint{}, nil
+		}
+
```
```suggestion
```

renaynay commented on das/store.go:
```diff
@@ -0,0 +1,114 @@
+	if err = s.Put(ctx, checkpointKey, bs); err != nil {
+		log.Errorw("storing checkpoint to disk", "Err", err)
+	}
+	log.Info("stored checkpoint to disk\n", cp.String())
```
```suggestion
	log.Infow("stored checkpoint to disk", "checkpoint", cp.String())
```

walldiss replied:
Actually it was this way, but then I've noticed that wrapped logger doesn't respect `Stringer` interface and just prints checkpoint as and object values instead.

renaynay commented on das/store.go:
```diff
@@ -0,0 +1,114 @@
+	}
+
+	if err = s.Put(ctx, checkpointKey, bs); err != nil {
+		log.Errorw("storing checkpoint to disk", "Err", err)
```
```suggestion
		log.Errorw("storing checkpoint to disk", "Err", err)
		return
```

otherwise will log success.

walldiss replied:
fixed

renaynay commented on das/store.go:
```diff
@@ -0,0 +1,114 @@
+	defer bgs.Done()
+
+	if storeInterval == 0 {
+		// run store routine only when storeInterval is non 0
```
Shouldn't this warn log out since it is unexpected?

walldiss replied:
I've kept it as a turn-off option for background store.

renaynay replied:
What does that mean?

walldiss replied:
backgroundStore could be disabled by passing storeInterval = 0. Adjusted comments

walldiss replied:
add log about the disabled background store

renaynay commented on das/store.go:
```diff
@@ -0,0 +1,114 @@
+
+		cp, err := bgs.getStats(ctx)
+		if err != nil {
+			continue
```
debug / warn log here?

walldiss replied:
added logs here

renaynay commented on das/store.go:
```diff
@@ -0,0 +1,114 @@
+
+type checkpointStore interface {
+	load(ctx context.Context) (checkpoint, error)
+	store(ctx context.Context, cp checkpoint)
```
why no error returned?

renaynay replied:
IMO this should still return an error and leave it up to caller to decide what to do with it.

renaynay commented on das/store.go:
```diff
@@ -0,0 +1,114 @@
+	log.Info("stored checkpoint to disk\n", cp.String())
+}
+
+func newBackgroundStore(store checkpointStore,
+	getStats func(ctx context.Context) (checkpoint, error)) backgroundStore {
```
```suggestion
func newBackgroundStore(
store checkpointStore,
	getStats func(ctx context.Context) (checkpoint, error),
	) backgroundStore {
```

and format

walldiss replied:
reformatted to multiline

Wondertan commented on das/store.go:
```diff
@@ -0,0 +1,114 @@
+		return
+	}
+	storeTicker := time.NewTicker(storeInterval)
+
```
Let's be nice and stop the ticker on return

walldiss replied:
fixed

renaynay reviewed (commented):
In the middle of reviewing, submitting comments so they don't get lost:

* godocs need attention
* why interface for checkpoint store?

renaynay commented on das/stats.go:
```diff
@@ -0,0 +1,30 @@
+	// CatchUpDone indicates whether all known headers are sampled
+	CatchUpDone bool `json:"catch_up_done"`
+	// IsRunning tracks whether routine is running
+	IsRunning bool `json:"is_running"`
```
This tracks only if the catch up process is running? We should make it explicit what `IsRunning` refers to. 

Also, the below statement will set IsRunning to `true` if the catchUp process is finished via `s.catchUpDone`. 
Only one of the below statements has to be true for IsRunning to be true. So if len(workers) == 0 BUT s.catchUpDone is true, then IsRunning will return true.
```
		IsRunning:     len(workers) > 0 || s.catchUpDone,
```

walldiss replied:
IsRunning is meant to indicate that DASer is running. If `catchUpDone == true && len(workers) == 0` DASer is still subscribed to new headers and can resume sampling once new header is found. Maybe add some clarity to comment line like:
"IsRunning tracks whether routine is running and ready to sample new headers"

renaynay replied:
Okay so `IsRunning tracks whether the DASer service is running'

walldiss replied:
rephrased the  comment

renaynay commented on das/subscriber.go:
```diff
@@ -0,0 +1,37 @@
+				return
+			}
+
+			log.Errorw("failed to get next header", "Err", err)
```
We should not just log and continue here because this could possibly cause a gap that is untracked by the daser, no?

walldiss replied:
DASer fills gaps when notified about new header

renaynay commented on das/subscriber.go:
```diff
@@ -0,0 +1,37 @@
+			log.Errorw("failed to get next header", "Err", err)
+			continue
+		}
+		log.Infow("found new header", "height", h.Height)
```
```suggestion
		log.Infow("new header received via subscription", "height", h.Height)
```

walldiss replied:
adjusted log msg, thank you!

renaynay replied:
it doesn't seem like it was fixed

walldiss replied:
fixed

renaynay commented on das/worker.go:
```diff
@@ -0,0 +1,93 @@
+	lock     sync.Mutex
+	state    workerState
+	resultCh chan<- result
+	sample   func(context.Context, uint64) error
```
why does this have to be stored on the worker? You can call `w.run()` and pass in the samplingCoordinaters sampleFn since it's the same for all sampling routines.

renaynay commented on das/worker.go:
```diff
@@ -0,0 +1,93 @@
+)
+
+type worker struct {
+	lock     sync.Mutex
```
you can just remove the `lock` alias and keep sync.Mutex directly on worker so you can do `w.Lock` instead of `w.lock.Lock`

renaynay replied:
nit but looks nicer imo

walldiss replied:
I usually prefer having it named so mutex public API doesn't accidentally get exported as a struct method.

renaynay reviewed (commented):
Another quick round: 

* when you receive header via subscription, you basically just get the height from it and then toss it, and then later inside the `nextJob` determination, ask header store to retrieve it for you. This is inefficient since the root is already given to you via the header in the subscription, but I understand why you do this in order to simplify the process. It's just not efficient to do additional I/O when you technically don't have to.

walldiss commented on das/stats.go:
```diff
@@ -0,0 +1,30 @@
+	// MaxKnown is the height of the newest known header
+	MaxKnown uint64 `json:"max_known_height"`
+	// Failed contains all skipped header's heights with corresponding try count
+	Failed map[uint64]int `json:"failed,omitempty"`
```
Some headers might have failed to be retrieved from header store (`GetByHeight` error), so it will require to store 2 Failed maps for retry:
   - failed to obtain from store  - stored by height
   - failed to be sampled - store by hash
   And it will make much harder retry logic.
Adding handling by hash would make coordinator much more complicated, right now it tracks all planned / in progress / failed headers by height. Internal hashes cache could do the job, but I thought that would be an overkill for now.

Wondertan replied:
I liked the idea of optimizing IO, but then I got a question: do we need to optimize IO for the failed cases here? Also, considering the complexity it introduces that Vlad described

renaynay commented on das/checkpoint.go:
```diff
@@ -0,0 +1,54 @@
+)
+
+type checkpoint struct {
+	SampledBefore uint64             `json:"sampled_before"`
+	MaxKnown      uint64             `json:"max_known"`
```
These two need better names since they are logged out. 
I suggest
`MaxKnown` => `LatestSampledHeader`
`SampledBefore` => `HeadOfSampledChain` (something to indicate that it is the tip of the contiguously successfully sampled chain of headers

walldiss replied:
Nice suggestion! Maybe do it a bit differently:
 - `MaxKnown` => `Head`: this header could be not sampled yet, so ` LatestSampledHeader`  could possible be misleading. On the other hand it is height of most recent Head so maybe just rename it to `Head`?
 - `SampledBefore` => `LatestSampledHeader` I think this is good name for it since all header before this one were sampled.

walldiss replied:
MaxKnown => NetworkHead
SampledBefore => HeadOfSampledChain in user interface
SampledBefore => SampleFrom in checkpoint

renaynay commented on das/doc.go:
```diff
@@ -13,10 +13,10 @@ sufficiently likely that all block data is available as it is assumed that there
-loop to DAS all headers between the checkpoint and the current network head. It simultaneously
+The central component of this package is the `samplingCoordinator`. It launches workers that
+perform DAS on new ExtendedHeaders in the network. The DASer kicks off this loop by loading
+its last DASed header (`checkpoint`) and kicking off worker pool
```
Can you also make a note about the priority queue?

walldiss replied:
Added line about recent headers prioritisation

renaynay commented on das/done.go:
```diff
@@ -0,0 +1,31 @@
+	}
+}
+
+func (sm *done) Done() {
```
Why is this exported?

walldiss replied:
API of `done` could be bnefitial outside of package. There is another topic that @Wondertan mentioned previously in review: this generic primitive should live in some utility package outside of DASer, we jsut need to find good spot for it

renaynay replied:
We generally should not export methods until they are actually useful outside the package.

Please no premature decisions like this, only leads to us asking "why is this thing exported again?" a couple months down the line

Wondertan replied:
There is no decision, and this is very minor. The public method on a private struct is the same as private, which is the case for the Done. This won't cause us any problems down the line.

renaynay reviewed (commented):
I am sorry for the incremental reviews 🙏🏻

distractedm1nd commented on das/store.go:
```diff
@@ -0,0 +1,112 @@
+// being able to store state on exit
+type backgroundStore struct {
+	store    *store
+	getStats func(ctx context.Context) (checkpoint, error)
```
Why is this abstracted when it is always the same function? If the issue is that it needs a handle to the coordinator, I don't understand why we can't implement the background store routine directly as a method of the coordinator (another question here: Why is the store in the coordinator, but the background store in the DASer?)

This would mean:
- There wouldn't be a need for a backgroundStore struct (assuming we also get rid of `done` here)
- `getCheckpoint` is called directly in the implementation of the `backgroundStore` routine, reducing abstraction and improving readability

walldiss replied:
I really don't want to put `backgroundStore` into `coordinator` because its functionality  is not related to `corrdinator's` purpose. `backgroundStore` is a separate routine and needs to  communicate state over `getstats` callback so there is no race conditions. Even tho it is same pkg and it is hard to follow dependancy graph  both `coordinator` and `backgroundStore` depend on `store` and not on each other. Thats why it should be controlled by DASer, not coordinator.

Wondertan replied:
Another option is to make a backreference. `backgroundStore` may keep the reference to the `coordinator` parent and call getStats directly from it. IMO, both ways are fine

walldiss replied:
pass funcs: store checkpointStore, getStats func(ctx context.Context) (checkpoint, error) to backgroundStore `run()` func

distractedm1nd commented on das/store.go:
```diff
@@ -0,0 +1,112 @@
+// The checkpoint store stores/loads the DASer's checkpoint to/from
+// disk using the checkpointKey. The checkpoint is stored as a struct
+// representation of the latest successfully DASed state.
+type store struct {
```
I am against creating an abstraction over datastore when you can just pass in a wrapped namespace to begin with, and use `datastore.Datastore` directly

Wondertan replied:
I think its fine in this case

distractedm1nd replied:
@Wondertan When I question an abstraction, it almost never has to do with performance.  It purely has to do with simplicity of the code - the abstraction here doesn't do anything helpful when it comes to understanding the code. The separation of concerns you mentioned was an argument I had about merging two structs that were abstractions over separate things - it does not apply here

distractedm1nd replied:
https://github.com/celestiaorg/celestia-node/issues/1049

distractedm1nd commented on das/worker.go:
```diff
@@ -0,0 +1,86 @@
+type job struct {
+	id         int
+	from, to   uint64
+	isPriority bool
```
How much of a performance boost will this bring/what are your arguments for why it is important? I am more on the side of keeping things as simple as they need to be, and I am unsure if this prioritization is worth the lines of code

walldiss replied:
Biggest benefit from introducing `priority` is the ability to prioritise recently found headers over aged ones. They are in much higher demand and needs to be propagated through network as fast as possible.

walldiss replied:
Main discussion: resolved
Sidenote: change json curr -> current

distractedm1nd replied:
also why have the struct name Curr? Especially if the json is now current

distractedm1nd commented on das/coordinator.go:
```diff
@@ -0,0 +1,139 @@
+	defer wg.Done()
+
+	select {
+	case sm.waitCh <- wg.Wait:
```
Just for understanding: So every time stats are collected (for example, for the background store), you stop spawning new workers until all jobs are completed, then you get the stats, and start spinning up workers again?

Wondertan replied:
It's a fancy way to lock the coordinator's routine and access its state. 
> Fancy because I haven't seen it before

distractedm1nd replied:
Fanciness or cleverness should be avoided if possible if it can be done in a simpler way - but that is a different topic

I still don't know the answer to my question, can you hold my hand? Does locking the coordinator's routine then mean that no new workers are spawned until all jobs are finished, so stats can be collected, and then the routine can continue?

walldiss replied:
My main goal was to make it as simple as possible tbh. 
Regarding first question - no. Workers are spawn until concurrencyLimit is reached. Then coordinator awaits for results from any worker before will spin out another one, if there are jobs available. In the meantime while workers are busy doing their jobs, corrdinator is not blocked from updating `maxKnown` or exporting stats to any external request.

This piece of code is responsible for preventing race condition on collecting stats from corrdinator. It blocks coordinator with waitGroup untill stats has been read and unblocks after by releasing waitGroup. This operation is also running on demand, while coordinator is not busy starting new workers or updating state.

Wondertan replied:
> Fanciness or cleverness should be avoided if possible if it can be done in a simpler way - but that is a different topic

Obviously, but any other method we can think of(mutexes or req/resp) does not apply in this case that well. It is not just _fancy_, it is also well suited

distractedm1nd reviewed (changes requested):
Not a full review (there is a lot for me to still read through, think about, and test), but I added some comments and questions with my current thoughts :)

In general I think things are maybe a bit too abstract and separated in places they don't need to be, and I still haven't understood completely how it works after looking at it for a few hours.

walldiss commented on das/store.go:
```diff
@@ -0,0 +1,112 @@
+
+	"github.com/ipfs/go-datastore"
+	"github.com/ipfs/go-datastore/namespace"
+	"golang.org/x/net/context"
```
wow, good catch! Thank you.

walldiss commented on das/daser.go:
```diff
@@ -17,49 +18,65 @@ import (
-	catchUpDn chan struct{} // done signal for catchUp loop
+	cancel         context.CancelFunc
+	subscriberDone chan struct{}
+	running        int32
```
It is part of 1.19 atomic package API. Unfortunately we are using 1.18 right nowt. There are many cool shiny things in new atomic update tho.

walldiss commented on das/subscriber.go:
```diff
@@ -0,0 +1,37 @@
+// subscriber subscribes for notifications about new headers in the network to keep
+// sampling process up-to-date with current network state.
+type subscriber struct {
+	done
```
subscriber runs the go-routine that subscribes to notifications about new headers in network. Upon founding new one it just triggers the callback to coordinator. So essentially even tho it is separate routine it has no access to shared state, nor any dependancy on it. Thats why it's for loop is so simple.

distractedm1nd replied:
No I understand this, sorry, abstraction was the wrong word (I was focusing on the struct itself). What I wanted to ask: How does the routine stop when the `done` struct signals it's finished?

walldiss replied:
It quits by `ctx` being canceled from outside

distractedm1nd replied:
Ok, makes sense, missed the if err == context.Canceled in the routine

walldiss commented on das/store.go:
```diff
@@ -0,0 +1,112 @@
+	defer ticker.Stop()
+
+	var prev uint64
+	for {
```
Not sure I understand question. Routine will run until `ctx` done channel is closed. Done struct is a done indicator for external check if routine has stopped already

Wondertan commented on das/coordinator.go:
```diff
@@ -0,0 +1,139 @@
+}
+
+// listen notifies coordinator about new header height has been found
+func (sm *samplingCoordinator) listen(ctx context.Context, height uint64) {
```
This could work, though, if the subscription were just a channel.

renaynay commented:
Okay so: 

My light node seemingly backwards "synced" the sample chain successfully. Now that I'm up to network head (in terms of sampling), I have the following output: 

```
{"sampled_before_height":194899,"max_known_height":194898,"concurrency":0,"catch_up_done":true,"is_running":true}
```

Why is sampled before height > max known height?

renaynay commented:
Another comment I have is that the `sampledBefore` metric is basically useless during the catch up process because the daser is performing a backwards "sync" of the chain when sampling. So `sampledBefore` is just perpetually = 2 since a worker samples from [1:1] on start, and then just starts prioritising new headers until it samples the entire chain from network head --> 2.

renaynay commented:
> Okay so:
> 
> My light node seemingly backwards "synced" the sample chain successfully. Now that I'm up to network head (in terms of sampling), I have the following output:
> 
> ```
> {"sampled_before_height":194899,"max_known_height":194898,"concurrency":0,"catch_up_done":true,"is_running":true}
> ```
> 
> Why is sampled before height > max known height?

`sampledBefore` is the "next" header to be sampled by the catchUp process correct? So technically once node is caught up, it sits at `maxKnown`+1 which can be confusing.

renaynay commented on das/state.go:
```diff
@@ -1,53 +1,234 @@
+	failed     map[uint64]int             // stores heights of failed headers with amount of attempt as value
+
+	nextJobID int
+	next      uint64 // all headers before next were sent to workers
```
should next not be protected somehow or read from / written to atomically?

renaynay replied:
same thing with `priority`

renaynay replied:
Unless it's guaranteed reads on `coordinatorState` can never happen at the same time as `nextJob` determination?

renaynay replied:
> corrdinator is not blocked from updating maxKnown or exporting stats to any external request.
This piece of code is responsible for preventing race condition on collecting stats from corrdinator. It blocks coordinator with waitGroup untill stats has been read and unblocks after by releasing waitGroup. This operation is also running on demand, while coordinator is not busy starting new workers or updating state.

`nextJob` will never be called simultaneously as stats are collected?

walldiss replied:
Yes. No need for protection, because it is protected by ownership scope.

renaynay reviewed (commented):
another round

walldiss commented on das/daser.go:
```diff
@@ -68,192 +88,59 @@ func (d *DASer) Start(context.Context) error {
+	cp, err := d.sampler.store.load(ctx)
 	if err != nil {
-		return err
+		h, err := d.getter.Head(ctx)
+		if err == nil {
+			cp = checkpoint{
+				SampledBefore: 1,
+				MaxKnown:      uint64(h.Height),
+			}
+
+			log.Warnw("checkpoint not found, initializing with height 1")
+		}
```
DASer tries to obtain Head info if icheckpoint is not available to start sampling process straight away. It also can get info about new head from `sub.NextHeader` later. I don't think stopping DASer completely on `Head` failure is necessarily if one of two sources  of head has been failed.

walldiss replied:
Add comment about error being non-critical for DASer init

walldiss commented on das/state.go:
```diff
@@ -1,53 +1,234 @@
+		st.priority = append(st.priority, st.newJob(wk.From, wk.To, true))
+	}
+
+	if c.SampledBefore == 0 {
```
It is possible if DASer was unable to load checkpoint from datastore and getting info from `Head()`

renaynay replied:
```go
	// load latest DASed checkpoint
	cp, err := d.sampler.store.load(ctx)
	if err != nil {
		log.Warnw("checkpoint not found, initializing with height 1")

		cp = checkpoint{
			SampledBefore: 1,
			MaxKnown:      1,
		}

		if h, err := d.getter.Head(ctx); err == nil {
			cp.MaxKnown = uint64(h.Height)
		}
	}
	log.Info("starting DASer from checkpoint:\n", cp.String())

	d.sampler.state = initCoordinatorState(samplingRange, cp)
```

If the DASer is unable to load the checkpoint, it sets the checkpoint `cp` to

```
checkpoint{
			SampledBefore: 1,
			MaxKnown:      1,
		}
```

tries to load the Head and set the MaxKnown to head.Height, then passes this checkpoint into `initCoordinatorState`.

So how can `SampledBefore` be 0 then unless the node can load a checkpoint from disk that has `SampledBefore stored as 0?

-----------------------

Another point, in this [commit](https://github.com/celestiaorg/celestia-node/pull/988/commits/b7979991f3628404f62d3bf740079c1d1e0aab3f#diff-359d277b92395b8825cbffc2b991baa1dc4e1a9092f94ab3519feb30bb8bcee8R101), you again ignore the error from getter.Head: 

```
		if h, err := d.getter.Head(ctx); err == nil {
			cp.MaxKnown = uint64(h.Height)
		}
```

Why do you ignore it? [Header store is initialised](https://github.com/celestiaorg/celestia-node/blob/main/node/components.go#L89) when the node is started up, where it checks to see if it can load a head, if not, it initialises store's Head to the header of the config's trusted Hash. So if you get an error here, it is an issue.

renaynay replied:
Sync discussion: Vlad is treating store kind of like a black box and doesn't "need" Head to operate the daser, which is why the error is ignored. We should probably doc it in the code though.

renaynay replied:
I still do not know how `SampledBefore` can be 0.

walldiss replied:
It can't. Previously it could be 0 between load and init funcs. I've moved setting default value for `SampledBefore` from `initCorrdinatorState` to `daser.Start` yesterday so `SampledBefore == 0` is impossible now

walldiss replied:
I'll add comment to describe `Head()` invocation as opportunistic mechanic

walldiss commented on das/state.go:
```diff
@@ -1,53 +1,234 @@
-type state struct {
-	sampleLk sync.RWMutex
-	sample   RoutineState // tracks information related to the main sampling routine
+// represents current state of sampling
```
done

walldiss commented on das/state.go:
```diff
@@ -1,53 +1,234 @@
+	catchUpDoneCh chan struct{} // blocks until all headers are sampled
+}
+
+func initSamplingState(samplingRangeSize uint64, c checkpoint) coordinatorState {
```
done

walldiss commented:
> Okay so:
> 
> My light node seemingly backwards "synced" the sample chain successfully. Now that I'm up to network head (in terms of sampling), I have the following output:
> 
> ```
> {"sampled_before_height":194899,"max_known_height":194898,"concurrency":0,"catch_up_done":true,"is_running":true}
> ```
> 
> Why is sampled before height > max known height?

 sampled_before_height is not included, only values that were before it. 
 Maybe address this by renaming variables suggestions discussion we had before?
 > SampledBefore => LatestSampledHeader I think this is good name for it since all header before this one were sampled.
 `latest_sampled_header` could be current `sampled_before_height` - 1
 Do you think it will be better this way?

walldiss commented:
I've addressed all open discussions implemented suggested solutions:
 -  Several renames and comments fixes. 
 -  Removed checkpoint store from coordinator, now checkpoint is stored by DASer itself.

Wondertan commented on das/daser.go:
```diff
@@ -140,37 +136,17 @@ func (d *DASer) Stop(ctx context.Context) error {
-	if err = d.cstore.store(ctx, newCheckpoint(d.sampler.state.unsafeStats())); err != nil {
-		log.Errorw("storing checkpoint to disk", "Err", err)
-	}
-
```
Why is this removed? The double logging issue should not be a reason for us to resample heights that were finished during `wait` stage.

walldiss replied:
returned it. I've suspected it could have had concurrency issues, but it is fine actualy.

Wondertan commented on das/daser.go:
```diff
@@ -22,7 +22,7 @@ var log = logging.Logger("das")
 const (
 	// samplingCoordinator will paginate headers into jobs with fixed size.
```
Pls, create an issue to make these params configurable

Wondertan commented on das/worker.go:
```diff
@@ -0,0 +1,97 @@
+	"go.uber.org/multierr"
+)
+
+type worker struct {
+	lock  sync.Mutex
+	state workerState
+}
+
+// workerState contains important information about the state of a
+// current sampling routine.
+type workerState struct {
+	job
+
+	Curr   uint64 `json:"current"`
+	Err    error  `json:"error,omitempty"`
+	failed []uint64
+}
+
+type job struct {
```
A bit off that only some structs are documented while others aren't

walldiss replied:
showed some love to other structs by adding comments

Wondertan commented on das/state.go:
```diff
@@ -1,53 +1,223 @@
+		failedFromWorker[h] = true
+	}
+
+	// update coordinator failed counter
+	for h := res.From; h <= res.To; h++ {
+		if failedFromWorker[h] {
+			// failed in worker, increase failed counter
+			s.failed[h]++
+			continue
+		}
+		delete(s.failed, h)
+	}
```
We should loop over `s.failed` instead. In this case, we loop over _each_ sampled header and delete for each, while, in fact, there will be almost none `s.failed` headers. This will save sooo many CPU cycles long run.

Something like this should work
```suggestion
	// check if the worker retried any of the previously failed heights
	for h, c := range s.failed {
		if h < res.From || h > rest.To {
			continue
		}
		
		if failedFromWorker[h] {
			s.failed[h]++
		} else {
			delete(s.failed, h)
		}
		delete(failedFromWorker, h)
	}
	// add newly failed heights
	for h := range failedFromWorker {
		s.failed[h]++
	}
```

Wondertan commented on das/state.go:
```diff
@@ -1,53 +1,223 @@
+	return true
+}
+
+// nextJob will return header height to be processed and done flog if there is none
```
```suggestion
// nextJob will return header height to be processed and done flag if there is none
```

Wondertan reviewed (commented):
The last set of nits and different findings before merging this.

Wondertan commented on das/worker.go:
```diff
@@ -0,0 +1,97 @@
+	log.Debugw("start sampling worker", "from", w.state.From, "to", w.state.To)
+
+	for curr := w.state.From; curr <= w.state.To; curr++ {
+		// TODO: get headers in batches
+		h, err := getter.GetByHeight(ctx, curr)
```
What is stopping us from getting ranges from getter now? If we want to keep it in another PR, let's make an issue.

distractedm1nd replied:
new PR please

walldiss replied:
There is an issue for that:
https://github.com/celestiaorg/celestia-node/issues/1015
Skipped it for now, because I need some time to dig deeper into `GetRangeByHeight` blocking mechanics.

distractedm1nd commented on das/state.go:
```diff
@@ -117,20 +116,13 @@ func (s *coordinatorState) nextJob() (next job, found bool) {
-		// skip job if already processed
-		if next.to <= s.next {
-			s.priority = s.priority[:len(s.priority)-1]
+		// skip the job, if it will be processed next without a priority
```
```suggestion
		// this job will be processed next normally, we can skip it
```

distractedm1nd commented on das/state.go:
```diff
@@ -117,20 +116,13 @@ func (s *coordinatorState) nextJob() (next job, found bool) {
 		}
 
-		// cut job if it is partially processed
-		if next.from <= s.next {
```
Why do we not need to do these checks anymore?

walldiss replied:
Thats a good question I should have answered in changelog of previous commit. Thank you for asking it!

distractedm1nd commented on das/stats.go:
```diff
@@ -5,10 +5,12 @@ package das
+	// all headers before HeadOfCatchup were submitted to sampling workers
+	HeadOfCatchup uint64 `json:"head_of_catchup"`
+	// NetworkHead is the height of the most recent header in the network
+	NetworkHead uint64 `json:"network_head_height"`
```
CatchupHead and SampledChainHead and more consistent with NetworkHead imo

walldiss replied:
renamed

distractedm1nd commented on das/stats.go:
```diff
@@ -17,12 +19,12 @@ type SamplingStats struct {
 	// CatchUpDone indicates whether all known headers are sampled
 	CatchUpDone bool `json:"catch_up_done"`
-	// IsRunning tracks whether routine is running
+	// IsRunning tracks whether the DASer service is running'
```
```suggestion
	// IsRunning tracks whether the DASer service is running
```

distractedm1nd commented on das/store.go:
```diff
@@ -25,18 +25,15 @@ type store struct {
 
-// loadCheckpoint loads the DAS checkpoint from disk and returns it.
-// If there is no known checkpoint, it returns a zero-value.
+// load the DAS checkpoint from disk and returns it.
```
```suggestion
// load loads the DAS checkpoint from disk and returns it.
```

distractedm1nd commented on das/state.go:
```diff
@@ -4,51 +4,45 @@ import (
-// represents current state of sampling
+// coordinatorState represents current state of sampling
 type coordinatorState struct {
 	rangeSize uint64
```
batchSize if you also like it more

distractedm1nd commented on das/state.go:
```diff
@@ -1,53 +1,223 @@
+}
+
+// initState initiates state for samplingCoordinator
+func initState(samplingRangeSize uint64) coordinatorState {
```
Can we name this newState or newCoordinatorState or something? init sounds like an action is being taken

distractedm1nd commented on das/store.go:
```diff
@@ -62,27 +59,26 @@ func (s *store) store(ctx context.Context, cp checkpoint) error {
+	return backgroundStore{newDone("background store")}
 }
 
 // run launches backgroundStore routine. Could be disabled by passing storeInterval = 0
```
```suggestion
// run launches the backgroundStore routine.
// the routine can be disabled by passing storeInterval = 0.
```

walldiss replied:
renamed

distractedm1nd commented on das/store.go:
```diff
@@ -0,0 +1,108 @@
+	checkpointKey = datastore.NewKey("checkpoint")
+)
+
+// The checkpoint store stores/loads the DASer's checkpoint to/from
+// disk using the checkpointKey. The checkpoint is stored as a struct
+// representation of the latest successfully DASed state.
+type store struct {
+	datastore.Datastore
+}
+
+// backgroundStore periodically saves current sampling state in case of DASer force quit before
+// being able to store state on exit
+type backgroundStore struct {
+	done
+}
```
This already has an open issue #1049

distractedm1nd replied:
But very very strongly agree.

Also, now the store is being passed into backgroundStore.run instead of being part of the struct? I don't understand this change

distractedm1nd reviewed (changes requested):
Not really a full review, I will need to come back to this and make some more comments I think

Please try to be more descriptive with your commits and add a comment if you fix any bugs /make any changes that don't fit in a commit message

walldiss commented on das/worker.go:
```diff
@@ -0,0 +1,97 @@
+	log.Debugw("start sampling worker", "from", w.state.From, "to", w.state.To)
+
+	for curr := w.state.From; curr <= w.state.To; curr++ {
+		// TODO: get headers in batches
+		h, err := getter.GetByHeight(ctx, curr)
+		if err == nil {
+			err = sample(ctx, h)
+		}
```
It is really hard to implement tests if ShareAvailable passed directly. Currently tests extensively check what header heights being sampled and in which order. ShareAvailable takes DataAvailabilityHeader as an input that has no easy way to map into header's height.
I would rather keep passing `sampleFn` as it allows dependency injection (like an interface, but without a struct receiver) on a level that has header height.

walldiss commented on das/daser.go:
```diff
@@ -37,6 +40,7 @@ type DASer struct {
 	bcast  fraud.Broadcaster
 	hsub   header.Subscriber // listens for new headers in the network
 	getter header.Getter     // retrieves past headers
+	cstore store
```
renamed

walldiss commented on das/daser.go:
```diff
@@ -63,12 +67,12 @@ func NewDASer(
 		bcast:          bcast,
 		hsub:           hsub,
 		getter:         getter,
+		cstore:         wrapCheckpointStore(dstore),
```
I think this is irrelevant after combining store and backgroundStore

walldiss commented on das/daser_test.go:
```diff
@@ -273,10 +274,36 @@ func (m *mockGetter) GetByHeight(_ context.Context, height uint64) (*header.Exte
 }
 
-func (m *mockGetter) GetRangeByHeight(ctx context.Context, from, to uint64) ([]*header.ExtendedHeader, error) {
+type benchGetterStab struct {
```
it is a typo haha

walldiss commented on das/worker.go:
```diff
@@ -0,0 +1,97 @@
+
+	"github.com/celestiaorg/celestia-node/header"
+
+	"go.uber.org/multierr"
```
fixed

walldiss commented on das/state.go:
```diff
@@ -1,53 +1,223 @@
+		failed:        make(map[uint64]int),
+		nextJobID:     0,
+		next:          1,
+		networkHead:   1,
```
done

walldiss commented on das/worker.go:
```diff
@@ -0,0 +1,97 @@
+	id   int
+	From uint64 `json:"from"`
+	To   uint64 `json:"to"`
+}
```
True. Removed redundant tags

walldiss commented on das/daser.go:
```diff
@@ -22,13 +22,16 @@ var log = logging.Logger("das")
 	backgroundStoreInterval = 10 * time.Minute
+
+	// priorityCap defines size limit of priority queue
+	priorityCap = defaultConcurrencyLimit * 4
```
+

walldiss commented on das/daser.go:
```diff
@@ -22,13 +22,16 @@ var log = logging.Logger("das")
 	// samplingCoordinator will paginate headers into jobs with fixed size.
 	//  samplingRange is the maximum amount of headers processed in one job.
-	samplingRange = 64
+	samplingRange = 100
```
My point for `range` is because that items are consecutive in between `from` and `to`. I feel like size would be nicer if it was a limit for amount of items in batch, but not necessarily having an order. 
Also header package has method `GetRangeByHeight` that will be introduced into DAS'er and will use this constant

walldiss commented on das/coordinator.go:
```diff
@@ -0,0 +1,129 @@
+	defer wg.Done()
+
+	select {
+	case sc.waitCh <- wg.Wait:
```
Thats another way of doing it. It could also be done with buffered channel, but implementation is more complicated. I feel like having a `Wait` is kinda helping to understand what is going on. I know this part is kinda confusing, so I've changed channel value type to full `*sync.WaitGroup` instead of  `func()` so there is less functional magic (but bigger interface passed).

walldiss commented:
I’ve squashed DAS’er commit history so it is easier to track new changes. All previous reviews comments were addressed before the squash.

Wondertan commented on das/daser.go:
```diff
@@ -68,268 +90,89 @@ func (d *DASer) Start(context.Context) error {
 			sendErr := d.bcast.Broadcast(ctx, fraud.CreateBadEncodingProof(h.Hash(), uint64(h.Height), byzantineErr))
 			if sendErr != nil {
-				log.Errorw("fraud proof propagating failed", "err", sendErr)
+				log.Errorw("fraud proof propagating failed", "Err", sendErr)
```
```suggestion
				log.Errorw("fraud proof propagating failed", "err", sendErr)
```

Is this a typo or did you intentionally make it capitalized?

Wondertan commented on das/store.go:
```diff
@@ -0,0 +1,101 @@
+func newCheckpointStore(ds datastore.Datastore) checkpointStore {
+	return checkpointStore{
+		namespace.Wrap(ds, storePrefix),
+		newDone("checkpoint checkpointStore")}
```
```suggestion
		newDone("checkpoint store")}
```

Wondertan commented on das/store.go:
```diff
@@ -0,0 +1,101 @@
+		}
+		if cp.SampleFrom > prev {
+			if err = s.store(ctx, cp); err != nil {
+				log.Errorw("storing checkpoint to disk", "Err", err)
```
```suggestion
				log.Errorw("storing checkpoint to disk", "err", err)
```

Wondertan commented on das/subscriber.go:
```diff
@@ -0,0 +1,37 @@
+				return
+			}
+
+			log.Errorw("failed to get next header", "Err", err)
```
```suggestion
			log.Errorw("failed to get next header", "err", err)
```

Seems like an issue from global renaming

Wondertan commented on das/worker.go:
```diff
@@ -0,0 +1,99 @@
+	defer w.lock.Unlock()
+	if err != nil {
+		w.state.failed = append(w.state.failed, curr)
+		w.state.Err = multierr.Append(w.state.Err, fmt.Errorf("height: %v, Err: %w", curr, err))
```
```suggestion
		w.state.Err = multierr.Append(w.state.Err, fmt.Errorf("height: %v, err: %w", curr, err))
```

Wondertan commented on das/daser.go:
```diff
@@ -68,268 +90,89 @@ func (d *DASer) Start(context.Context) error {
 		log.Errorw("sampling failed", "height", h.Height, "hash", h.Hash(),
-			"square width", len(h.DAH.RowsRoots), "data root", h.DAH.Hash(), "err", err)
-		// report previous height as the last successfully sampled height
+			"square width", len(h.DAH.RowsRoots), "data root", h.DAH.Hash(), "Err", err)
```
```suggestion
			"square width", len(h.DAH.RowsRoots), "data root", h.DAH.Hash(), "err", err)
```

Wondertan commented on das/daser.go:
```diff
@@ -68,268 +90,89 @@ func (d *DASer) Start(context.Context) error {
-			}
-		}
+	if err = d.store.store(ctx, cp); err != nil {
+		log.Errorw("storing checkpoint to disk", "Err", err)
```
```suggestion
		log.Errorw("storing checkpoint to disk", "err", err)
```

Wondertan commented on das/store.go:
```diff
@@ -0,0 +1,101 @@
+
+	var prev uint64
+	for {
+		// blocked by ticker to perform checkpointStore only once in a period of time
```
```suggestion
		// blocked by ticker to perform storing only once in a period
```

Wondertan reviewed (dismissed):
Just comments and logs fixes. The logic is correct, so a solid approval.

Wondertan reviewed (approved):
Reapproval

renaynay commented on das/state.go:
```diff
@@ -1,53 +1,228 @@
-	Start time.Time `json:"start"`
-	End   time.Time `json:"end"`
-	Error error     `json:"error"`
+func (s *coordinatorState) updateHead(last uint64) bool {
```
why does this return a bool if it's never checked?

walldiss replied:
I forgot to remove it after branching metrics into separate PR. It will be used in metrics PR to trigger new_head metric.

renaynay reviewed (commented):
Awesome progress on this PR, it looks great now and much easier to follow :) thanks a lot for being so thorough with this.

Nit that can be left for a subsequent PR: 
we should godoc every exported function, struct etc. (e.g. `WorkerStats`) even if it seems obvious as it's best practice

walldiss commented on das/done.go:
```diff
@@ -0,0 +1,31 @@
+	}
+}
+
+func (sm *done) indicateDone() {
```
Thank you, renamed

walldiss commented on das/daser.go:
```diff
@@ -4,62 +4,84 @@ import (
 
+// TODO: parameters needs performance testing on real network to define optimal values
+const (
+	// samplingCoordinator will paginate headers into jobs with fixed size.
```
Removed

renaynay reviewed (approved):
morge it

Wondertan commented:
Let's also resolve all the comments to be nice and tidy
